src/EoS/Nuclear/findtoreps_direct.cpp

```diff
@@ -0,0 +1,191 @@
+//                    Entropy     mode (2)
+//                    Pressure    mode (3)
+//
+// Note        :  1. Use 3D Lagrange linear interpolation or Catmull-Rom cubic interpolation formulae
+//                   to search the corresponding temperature/energy given (rho, eps/temp/entr/pres, Ye)
+//                2. Invoked by nuc_eos_C_short()
+//
+// Parameter   :  x             : Input vector of first  variable (rho)
+//                var           : Input vector of second variable (eps/temp/entr/pres)
+//                z             : Input vector of third  variable (Ye)
+//                found_ltoreps : Output log(temp)/log(eps) of interpolated function values
+//                table_main    : The main      tabulated nuclear EoS table
+//                table_Aux     : The auxiliary tabulated nuclear EoS table
+//                nx            : X-dimension of table
+//                ny            : Y-dimension of table
+//                nz            : Z-dimension of table
+//                nvar          : Size of input quantity in the auxiliary table
+//                xt            : Vector of x-coordinates of table
+//                yt            : Vector of y-coordinates of table
+//                zt            : Vector of z-coordinates of table
+//                vart          : Vector of the input quantity
+//                keymode       : Which mode to be used
+//                keyerr        : Output error
+//
+// Return      :  found_ltoreps
+//-------------------------------------------------------------------------------------
+GPU_DEVICE
+void findtoreps_direct( const real x, const real var, const real z,
+                        real *found_ltoreps, const real *table_main, const real *table_aux,
+                        const int nx, const int ny, const int nz, const int nvar,
+                        const real *xt, const real *yt, const real *zt, const real *vart,
+                        const int keymode, int *keyerr )
+{
+
+   int  ix, iy_brd1, iy_brd2, iz, iv, ip, ifs;
+   real dx, dz, dv, dxi, dzi, dvi, fracx, fracy, fracz;
+   real var_brd1, var_brd2;
+
+   const int nxy  = nx  * ny;
+   const int nxyz = nxy * nz;
+
+
+// determine spacing parameters and location in an equidistant table
+   dx = (   xt[nx  -1] -   xt[0] ) / (real)(nx  -1);  // dens
+   dz = (   zt[nz  -1] -   zt[0] ) / (real)(nz  -1);  // ye
+   dv = ( vart[nvar-1] - vart[0] ) / (real)(nvar-1);  // var
+
+   dxi = (real)1.0 / dx;
+   dzi = (real)1.0 / dz;
+   dvi = (real)1.0 / dv;
+
+   ix = (int)(  (   x -   xt[0] ) * dxi  );
+   iz = (int)(  (   z -   zt[0] ) * dzi  );
+   iv = (int)(  ( var - vart[0] ) * dvi  );  // index in the auxiliary table
+
+
+// (0) check if the target variable is in the degeneracy region
+   const int iv_left  =     (int)table_aux[ix + nx *  iv    + nxy * iz];  // index in the main table
+   const int iv_right = 1 + (int)table_aux[ix + nx * (iv+1) + nxy * iz];  // index in the main table
```

Comment:
I think the comment is incorrect. It might be auxiliary table.

Reply:
No. `iv_left` and `iv_right` are indices in the main table. For the direct method, the auxiliary table stores the index `iy_bdry_1`such that `var(ix, iy_bdry_1, iz) <= var < var(ix, iy_bdry_1+1, iz)`, for given dens, Ye, and `var`. And use `-1` to indicate whether the given `var` is in the degenerate region.

Reply:
I see. I misunderstood. Thanks for the explanation!